### PR TITLE
Fix transform incorrectly calculating date bucket on updating old data 

### LINKED ISSUE #97101: Transform incorrectly calculates date bucket after updating older data
### Elasticsearch Version

8.6.2

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

Linux 20.04.2-Ubuntu

### Problem Description

Continuous tranform giving invalid result when date_histogram is used in pivot 

### Steps to Reproduce

// Step 1 Add ingest pipeline
PUT /_ingest/pipeline/es-timeadd_1?pretty
{
  "processors": [
    {
      "set": {
        "field": "_source.es_timestamp",
        "value": "{{_ingest.timestamp}}"
      }
    }
  ]
}

// Step2 Create index using above pipeline
PUT /test
{
  "settings": {
    "index.default_pipeline" : "es-timeadd"
  }
}

// Step3 Insert two documents
PUT /test/_doc/1
{
  "id" : "ritesh",
  "gen_timestamp_min" : "2023-06-21T15:54:00.000Z",
  "flow_count" : 5
}

PUT /test/_doc/2
{
  "id" : "ritesh",
  "gen_timestamp_min" : "2023-06-21T15:55:00.000Z",
  "flow_count" : 5
}

// Step4 - Create transform
PUT /_transform/test-transform
{
  "source": {
    "index": "test"
  },
  "dest" : {
    "index" : "test-transformed-index"
  },
  "sync": {
    "time": {
      "field": "es_timestamp",
      "delay": "0s"
    }
  },
  "pivot": {
    "group_by": {
      "gen_timestamp_hour": {
        "date_histogram": {
          "field": "gen_timestamp_min",
          "fixed_interval": "1h"
        }
      },
      "asset": {
        "terms": {
          "field": "id.keyword"
        }
      }
    },
    "aggregations": {
      "flow_count": {
        "sum": {
          "field": "flow_count"
        }
      }
    }
  }
}

// Step 5 - Start Transform
POST /_transform/test-transform/_start

// Wait for few seconds
// Step6 - Check data in destination index
GET test-transformed-index/_search
{
}

We get below output, In which sum of flow counts is 10 which is valid
{
  "took": 0,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test-transformed-index",
        "_id": "cgAnWNlrvx_jB6WvSZa9f0n5AAAAAAAA",
        "_score": 1,
        "_source": {
          "gen_timestamp_hour": "2023-06-21T15:00:00.000Z",
          "flow_countNAME": 10,
          "asset": "ritesh"
        }
      }
    ]
  }
}

// Step 7 Update document in source index with id 1, as below
PUT /test/_doc/1
{
  "id" : "ritesh",
  "gen_timestamp_min" : "2023-06-21T15:54:00.000Z",
  "flow_count" : 6
}

// Step 8 Check data in output index, Total flow count should be 11, instead we see only 6 as below
GET test-transformed-index/_search
{
  
}

{
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "test-transformed-index",
        "_type" : "_doc",
        "_id" : "cgAnWNlrvx_jB6WvSZa9f0n5AAAAAAAA",
        "_score" : 1.0,
        "_source" : {
          "gen_timestamp_hour" : "2023-06-21T15:00:00.000Z",
          "flow_countNAME" : 6.0,
          "asset" : "ritesh"
        }
      }
    ]
  }
}

Root Cause - Which reprocessing in continuous tranform, It picks data upto gen_timestamp_min of modified document, It should process all the data in given group by.

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by alyokaz**
This pull request fixes #97101, where a continuous pivot transform was incorrectly calculating a metric aggregation on a date bucket when a document from that bucket was updated. This was due to the transform taking the updated document's timestamp field as an upper bound in the transforms optimisation query and, consequently, ignoring all documents with a later timestamp in the updated metric aggregation calculation for the bucket.

I changed the the rounding of `upperBound` in `DateHistogramFieldCollector` to round up to the next time interval, as this is where the transform's optimisation query pulls its upper bound from. This will have the effect of all the documents in the same bucket as the document being updated being pulled into the re-calculation of the buckets aggregations.

I've updated and refactored out the unit tests from `CompositeBucketsChangeCollectorTests` into a new class `DateHistorgramFieldCollectorTests`. This was done in the process of reading and understanding the tests. I'm not sure if the refactor is necessary or adds unnecessary bloat, but the same could be done for the other `FieldCollector` tests. It has been added as a separate commit as advised. 

I've added an integration test in `TransformPivotRestIT`. I chose this rather then `TransformContinuousIT` as the specific set up needed to recreate the issue didn't seem to fit well with the approach used there. The test could be refactored / integrated into the test harness further, however, I thought I'd leave it as is just to be more explicit about what was done.

### CODE REVIEW

File: x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        new DateHistogramGroupSource.FixedInterval(DateHistogramInterval.MINUTE),
+        null,
+        null
+    );
+
+    final double MIN_TIMESTAMP_VALUE = 122_633;
+    final double MAX_TIMESTAMP_VALUE = 302_525;
+    final double EXPECTED_LOWER_BOUND = 120_000;
+
+    final double EXPECTED_UPPER_BOUND = 360_000;
```
Comment (hendrikmuhs) [inline x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java:59]: constants should be `private static final`

File: x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+import static org.hamcrest.CoreMatchers.instanceOf;
+import static org.hamcrest.Matchers.equalTo;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class DateHistorgramFieldCollectorTests extends ESTestCase {
+    Map<String, SingleGroupSource> groups = new LinkedHashMap<>();
+
+    SingleValue minTimestamp = mock(NumericMetricsAggregation.SingleValue.class);
+    SingleValue maxTimestamp = mock(NumericMetricsAggregation.SingleValue.class);
```
Comment (hendrikmuhs) [inline x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java:40]: a mix of per suite and per test (`@Before`) initialization is unclean. Apart from that, members should be `private`. I think it is cleaner if every test is independent and uses own mocks. You can still re-use code. My suggestion is to introduce but not initialize the variable:

```
private SingleValue minTimestamp;
private SingleValue maxTimestamp;
```

and set it up in 

```
    @Before
    public void setupDateHistogramFieldCollectorTest() {
         minTimestamp = mock(NumericMetricsAggregation.SingleValue.class);
         maxTimestamp = mock(NumericMetricsAggregation.SingleValue.class);
        ...
```

File: x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    );
+
+    final double MIN_TIMESTAMP_VALUE = 122_633;
+    final double MAX_TIMESTAMP_VALUE = 302_525;
+    final double EXPECTED_LOWER_BOUND = 120_000;
+
+    final double EXPECTED_UPPER_BOUND = 360_000;
+
+    @Before
+    public void setUpDateHistogramFieldCollectorTests() {
```
Comment (hendrikmuhs) [inline x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java:62]: `setUp` -> `setup` (one word). `@Before` works per test, so `setupDateHistogramFieldCollectorTest` not `setupDateHistogramFieldCollectorTests`

File: x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        assertNull(queryBuilder);
+
+        // field and sync_field are different
+        collector = CompositeBucketsChangeCollector.buildChangeCollector(groups, SYNC_TIMESTAMP);
+        queryBuilder = buildFilterQuery(collector);
+
+        assertNull(queryBuilder);
+    }
+
+    private void assertQuery(QueryBuilder queryBuilder, Double expectedLowerBound, Double expectedUpperBound, String expectedFieldName) {
```
Comment (hendrikmuhs) [inline x-pack/plugin/transform/src/test/java/org/elasticsearch/xpack/transform/transforms/pivot/DateHistorgramFieldCollectorTests.java:128]: this and the below owns should be static I think

**Comment by hendrikmuhs** [comment]
@elasticmachine test this please

**Review by hendrikmuhs — COMMENTED** [review]
Thanks for your contribution, I've triggered test execution in CI and did a 1st quick review and added some comments.

**Comment by alyokaz** [comment]
I've made the constants private and static, moved the initialisation of the `minTimestamp`, `maxTimestamp` and `groups` to `setupDateHistogramFieldCollectorTest`, made the utility / pure functions static and fixed the typos as suggested.

**Comment by hendrikmuhs** [comment]
@elasticmachine test this please

**Comment by hendrikmuhs** [comment]
Please address the checkstyle issue, you can `gradlew precommit` locally, the output tells you how to format the code easily with another gradle command.

Please also merge main into the branch(feel free to use `rebase` to keep your commits on top).

The PR looks almost finished, good work!

**Comment by alyokaz** [comment]
I've fixed the checkstyle violations and rebased the branch on main.

Thanks, glad it's up to scratch.

**Comment by hendrikmuhs** [comment]
@elasticsearchmachine generate changelog

**Comment by hendrikmuhs** [comment]
@elasticmachine test this please

**Comment by hendrikmuhs** [comment]
@elasticmachine generate changelog

**Comment by hendrikmuhs** [comment]
@elasticmachine test this please

**Review by hendrikmuhs — APPROVED** [review]
LGTM

Looks great! Thank you for your contribution.

**Comment by alyokaz** [comment]
Happy to contribute, thanks for the review.

**Pull request merged**